dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java

```diff
@@ -0,0 +1,256 @@
+import io.openliberty.netty.internal.BootstrapExtended;
+import io.openliberty.netty.internal.ConfigConstants;
+import io.openliberty.netty.internal.exception.NettyException;
+import io.openliberty.netty.internal.impl.NettyFrameworkImpl;
+import io.openliberty.netty.internal.impl.NettyConstants;
+
+public class UDPUtils {
+
+    private static final TraceComponent tc = Tr.register(UDPUtils.class, UDPMessageConstants.NETTY_TRACE_NAME,
+            UDPMessageConstants.UDP_BUNDLE);
+
+    /**
+     * Create a {@link BootstrapExtended} for inbound UDP channels
+     * 
+     * @param framework
+     * @param options
+     * @return
+     * @throws NettyException
+     */
+    public static BootstrapExtended createUDPBootstrap(NettyFrameworkImpl framework, Map<String, Object> options)
+            throws NettyException {
+        return create(framework, new UDPConfigurationImpl(options, true));
+    }
+
+    /**
+     * Create a {@link BootstrapExtended} for outbound channels
+     * 
+     * @param framework
+     * @param options
+     * @return
+     * @throws NettyException
+     */
+    public static BootstrapExtended createUDPBootstrapOutbound(NettyFrameworkImpl framework,
+            Map<String, Object> options) throws NettyException {
+        return create(framework, new UDPConfigurationImpl(options, false));
+    }
+
+    private static BootstrapExtended create(NettyFrameworkImpl framework, UDPConfigurationImpl config) {
+        BootstrapExtended bs = new BootstrapExtended();
+        bs.applyConfiguration(config);
+        bs.group(framework.getParentGroup());
+        bs.channel(NioDatagramChannel.class);
+        return bs;
+    }
+
+    private static ChannelFuture bind(NettyFrameworkImpl framework, BootstrapExtended bootstrap, String inetHost,
+            int inetPort, final int retryCount, final int retryDelay, ChannelFutureListener bindListener) {
+        ChannelFuture bindFuture = bootstrap.bind(inetHost, inetPort);
+        if (inetHost.equals("*")) {
+            inetHost = NettyConstants.INADDR_ANY;
+        }
+        final String newHost = inetHost;
+
+        if (bindListener != null) {
+            bindFuture.addListener(bindListener);
+        }
+        final UDPConfigurationImpl config = ((UDPConfigurationImpl) bootstrap.getConfiguration());
+        final String channelName = ((UDPConfigurationImpl) bootstrap.getConfiguration()).getExternalName();
+
+        bindFuture.addListener(future -> {
+            if (future.isSuccess()) {
+
+                // add the new channel to the set of active channels, and set a close future to
+                // remove it
+                final Channel channel = bindFuture.channel();
+                framework.getActiveChannels().add(bindFuture.channel());
+                channel.closeFuture().addListener(innerFuture -> framework.stop(channel));
+
+                // set common channel attrs
+                channel.attr(ConfigConstants.NAME_KEY).set(config.getExternalName());
```

Comment:
You should set the state of channel BEFORE setting up a listener, 99.999% of time (e.g. during testing) the stop method with be able to use/trace the name, host and port and then one day it won't.

Reply:
Understood, updated to add the listener after setting comment attributes